File: week10/ch.Choe/Week10/Week10/SoldState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Week10
{
    class SoldState : State
    {
        //GumballMachine m_GumballMachine;

        public static SoldState(GumballMachine machine)
        {
            //m_GumballMachine = machine;
        }
        public static void Dispense(machine m)
        {
            Console.WriteLine("알맹이가 나가고 있습니다.");
            m.ReleaseBall();
            if(m_GumballMachine.Getcount() > 0)
            {
                m.SetState(GumballMachine.StateType.NoQuarter);
            }
            else
            {
                m.SetState(GumballMachine.StateType.Soldout);
            }
        }

        public static void EjectQuarter()
        {
            Console.WriteLine("이미 뽑았습니다.");
        }

        public static void InsertQuarter()
        {
            Console.WriteLine("잠깐 기다려 주세요.");
        }

        public static void TurnCrack()
        {
            Console.WriteLine("손잡이는 한번만.");
        }
    }
}

File: week10/younggeun0/state/SoldState.java
package week10.younggeun0.state;

public class SoldState implements State {
	GumballMachine2 gumballMachine;
	
	public SoldState(GumballMachine2 gumballMachine) {
		this.gumballMachine = gumballMachine;
	}
	@Override
	public void insertQuarter() {
		System.out.println("잠깐만 기다려 주세요. 알맹이가 나가고 있습니다");
	}
	@Override
	public void ejectQuarter() {
		System.out.println("이미 알맹이를 뽑으셨습니다.");
	}
	@Override
	public void turnCrank() {
		System.out.println("손잡이는 한 번만 돌려주세요.");
	}
	@Override
	public void dispense() {
		gumballMachine.releaseBall();
		if (gumballMachine.getCount() > 0) {
			gumballMachine.setState(gumballMachine.getNoQuarterState());
		} else {
			System.out.println("Opps, 다 팔렸네");
			gumballMachine.setState(gumballMachine.getSoldOutState());
		}
	}
}
